server/src/main/scala/org/http4s/server/middleware/Throttle.scala

```diff
@@ -0,0 +1,121 @@
+package org.http4s.server.middleware
+
+import org.http4s.{Http, Response, Status}
+import cats.data.Kleisli
+import cats.effect.{Clock, Sync}
+import cats.effect.concurrent.Ref
+import scala.concurrent.duration.FiniteDuration
+import cats.implicits._
+import java.util.concurrent.TimeUnit.NANOSECONDS
+import scala.concurrent.duration._
+
+/**
+  * Transform a service to reject any calls the go over a given rate.
+  */
+object Throttle {
+
+  sealed abstract class TokenAvailability extends Product with Serializable
+  case object TokenAvailable extends TokenAvailability
+  final case class TokenUnavailable(retryAfter: Option[FiniteDuration]) extends TokenAvailability
+
+  /**
+    * A token bucket for use with the [[Throttle]] middleware.  Consumers can take tokens which will be refilled over time.
+    * Implementations are required to provide their own refill mechanism.
+    *
+    * Possible implementations include a remote TokenBucket service to coordinate between different application instances.
+    */
+  trait TokenBucket[F[_]] {
+    def takeToken: F[TokenAvailability]
+  }
+
+  object TokenBucket {
+
+    /**
+      * Creates an in-memory [[TokenBucket]].
+      *
+      * @param capacity the number of tokens the bucket can hold and starts with.
+      * @param refillEvery the frequency with which to add another token if there is capacity spare.
+      * @return A task to create the [[TokenBucket]].
+      */
+    def local[F[_]](capacity: Int, refillEvery: FiniteDuration)(
+        implicit F: Sync[F],
+        clock: Clock[F]): F[TokenBucket[F]] = {
+
+      def getTime = clock.monotonic(NANOSECONDS)
+      val bucket = getTime.flatMap(time => Ref[F].of((capacity.toDouble, time)))
+
+      bucket.map { counter =>
+        new TokenBucket[F] {
+          override def takeToken: F[TokenAvailability] = {
+            val attemptUpdate = counter.access.flatMap {
+              case ((previousTokens, previousTime), setter) =>
+                getTime.flatMap(currentTime => {
+                  val timeDifference = currentTime - previousTime
+                  val tokensToAdd = timeDifference.toDouble / refillEvery.toNanos.toDouble
+                  val newTokenTotal = Math.min(previousTokens + tokensToAdd, capacity.toDouble)
+
+                  val attemptSet: F[Option[TokenAvailability]] = if (newTokenTotal >= 1) {
+                    setter((newTokenTotal - 1, currentTime))
+                      .map(_.guard[Option].as(TokenAvailable))
+                  } else {
+                    val timeToNextToken = refillEvery.toNanos - timeDifference
+                    val successResponse = TokenUnavailable(timeToNextToken.nanos.some)
+                    setter((newTokenTotal, currentTime)).map(_.guard[Option].as(successResponse))
```

Comment:
Just to double check: Is this intended that if you keep sending requests while being throttled, it will remain throttled? I.e: If this is the case and you hit a throttle, while continuing to send requests, it will not let any through after the throttle, since it always updated the current time here.

Reply:
That is not the intention, and I don't *think* that is the behaviour here, though my reasoning could be wrong... it should keep adding fractional tokens until there is a whole token at which point the request passes, if you're unconvinced then I can add a test that with (hopefully!) prove this?

Reply:
Yeah I was wrong, didn't take into account fractional updates.